tango/integrations/wandb/step_cache.py

```diff
@@ -0,0 +1,227 @@
+import logging
+import os
+import shutil
+import tempfile
+from pathlib import Path
+from typing import Any, Optional, Union
+
+import wandb
+from wandb.errors import Error as WandbError
+
+from tango.common.aliases import PathOrStr
+from tango.common.file_lock import FileLock
+from tango.common.params import Params
+from tango.common.util import tango_cache_dir
+from tango.step import Step
+from tango.step_cache import CacheMetadata, StepCache
+from tango.step_caches.local_step_cache import LocalStepCache
+from tango.step_info import StepInfo
+
+from .util import ArtifactKind, check_environment, is_missing_artifact_error
+
+logger = logging.getLogger(__name__)
+
+
+# This class inherits from `LocalStepCache` to benefit from its in-memory "weak cache" and "strong cache",
+# but it handles saving artifacts to disk a little differently.
+@StepCache.register("wandb")
+class WandbStepCache(LocalStepCache):
+    """
+    This is a :class:`~tango.step_cache.StepCache` that's used by :class:`WandbWorkspace`.
+    It stores the results of steps on W&B as Artifacts.
+
+    It also keeps a limited in-memory cache as well as a local backup on disk, so fetching a
+    step's resulting subsequent times should be fast.
+
+    :param project: The W&B project to use.
+    :param entity: The W&B entity (user or organization account) to use.
+
+    .. tip::
+        Registered as :class:`~tango.step_cache.StepCache` under the name "wandb".
+    """
+
+    def __init__(self, project: str, entity: str):
+        check_environment()
+        super().__init__(tango_cache_dir() / "wandb_cache")
+        self.project = project
+        self.entity = entity
+
+    @property
+    def wandb_client(self) -> wandb.Api:
+        return wandb.Api(overrides={"entity": self.entity, "project": self.project})
+
+    def _acquire_step_lock_file(self, step: Union[Step, StepInfo], read_only_ok: bool = False):
+        return FileLock(
+            self.step_dir(step).with_suffix(".lock"), read_only_ok=read_only_ok
+        ).acquire_with_updates(desc=f"acquiring step cache lock for '{step.unique_id}'")
```

Comment:
I think the reason `LocalWorkspace` does that is because it's built so that it can be running multiple steps at the same time from the same `LocalWorkspace` instance. But that's not possible at the moment with `WandbWorkspace`, since you can only have one W&B run in a process at once.